Implement amp-addthis extension

jmrog opened the pull request:
# Implement amp-addthis extension

This PR implements the amp-addthis extension described in Issue #6360. The extension renders widgets -- sharing widgets, for now, but eventually (we hope) "follow" widgets, "related posts" widgets, etc. -- configured via the AddThis publisher dashboard at https://www.addthis.com/dashboard. Features of the extension currently include:

- Support for around 200 sharing services
- Support for analytics data including per-service and total share counts, view counts, most-shared content, etc. (available via the AddThis.com dashboard, but also (in the case of share counts) displayed on the widgets themselves)
- Extensive customization options
- HTTP response caching (to prevent unnecessary duplicate requests to the addthis.com dashboard when multiple amp-addthis widgets are included on a page)
- Use of the "light" version of the AddThis JS client described in #6360

AddThis' parent company, Oracle, has signed the corporate CLA and approved AddThis' contribution of this extension. Thanks!

Related-to #6360.

jridgewell commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,190 @@
+      'background-color': '#eee',
+    });
+
+    const image = win.document.createElement('amp-img');
```
Nit: we have a [createElementWithAttributes](https://github.com/ampproject/amphtml/blob/3b60020ca0478d83289458c2af7b92da8b8918db/src/dom.js#L162-L172) that could clean this up.

jmrog replied:
Oh, nice! Thanks; I'll update the diff as soon as I can.

jridgewell commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,190 @@
+
+    const placeholder = win.document.createElement('div');
+    placeholder.setAttribute('placeholder', '');
+    setStyles(placeholder, {
```
Nit: `setStyle` for single style.

jridgewell commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,190 @@
+    );
+  }
+
+  createPlaceholderCallback() {
```
Woo!

jridgewell commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,190 @@
+
+  /** @override */
+  layoutCallback() {
+    const iframe = this.element.ownerDocument.createElement('iframe');
```
Nit: `createElementWithAttributes`

jridgewell commented on extensions/amp-addthis/0.1/ConfigManager.js:
```diff
@@ -0,0 +1,244 @@
+    }
+
+    if (!this.ownerDocument_) {
+      this.ownerDocument_ = element.ownerDocument;
```
This will have issues for embedding PWA, etc.

jmrog replied:
@jridgewell Do you think you could say more about the issue here? For what it's worth, `element` in this case is just the AmpElement's `this.element`, and the intent is to get the title (etc.) set on the AmpElement's document using it. I could just pass each particular value that I'm interested in, instead, if the concern is an additional reference to the element's `ownerDocument`. Or is the concern that these values wouldn't always be accurate when the element is deeply embedded (or won't necessarily reflect the app's name, etc., in the case of a PWA)? Or something else?

Thanks.

jridgewell replied:
> Or is the concern that these values wouldn't always be accurate when the element is deeply embedded

Exactly.

jridgewell commented on extensions/amp-addthis/0.1/ConfigManager.js:
```diff
@@ -0,0 +1,244 @@
+    const pubData = this.dataForPubId_[pubId];
+    const dashboardConfig = pubData.config;
+    const configRequestStatus = pubData.requestStatus;
+    const widgetId = iframe[WIDGET_ID_PROP];
```
It looks like this could just be passed in instead of mutating a DOM element object.

honeybadgerdontcare commented:
@jmrog you may want to [read this](https://github.com/ampproject/amphtml/blob/master/contributing/component-validator-rules.md) about creating validator rules for your component.

jmrog commented:
@jridgewell @honeybadgerdontcare Okay, thanks! I actually _intentionally_ didn't create validator rules due to the language in the documentation [here](https://github.com/ampproject/amphtml/blob/master/contributing/building-an-amp-extension.md#allowing-proper-validations) (specifically, the bit saying "you _need_ to file an issue on the GitHub repo select "Related to: Validator'" (emphasis added)). I now see that the longer document about validator rules actually says that this is optional, though, so I'm happy to work on adding validator rules to this PR.

honeybadgerdontcare commented:
@ampproject/validator Thanks for pointing that out, we should update that document to clarify things.

bpaduch commented on extensions/amp-addthis/amp-addthis.md:
```diff
@@ -0,0 +1,74 @@
+<table>
+  <tr>
+    <td width="40%"><strong>Description</strong></td>
+    <td>Displays an AddThis website tools embed.</td>
```
Revise this to include a link to AddThis site:

`Displays an <a href="https://www.addthis.com">AddThis</a> website tools embed.`

bpaduch commented on extensions/amp-addthis/amp-addthis.md:
```diff
@@ -0,0 +1,74 @@
+    <td class="col-fourty"><strong><a href="https://www.ampproject.org/docs/guides/responsive/control_layout.html">Supported Layouts</a></strong></td>
+    <td>fill, fixed, fixed-height, flex-item, nodisplay, responsive</td>
+  </tr>
+  <tr>
```
Remove this row, if there is no examples that are linkable.

bpaduch commented on extensions/amp-addthis/amp-addthis.md:
```diff
@@ -0,0 +1,74 @@
+</table>
+
+## Behavior
+
```
Add a statement about this extension, for example:

The `amp-addthis` component provides the ability to embed <a href="https://www.addthis.com">AddThis</a> website tools (for example, sharing buttons, follow buttons, related posts, etc.) on an AMP page.

bpaduch commented on extensions/amp-addthis/amp-addthis.md:
```diff
@@ -0,0 +1,74 @@
+## Validation
+
+<!---
+TODO: Restore this once validation is completed.
```
Please ensure this section is completed when the validation is implemented, using the phrasing:

See amp-addthis rules in the AMP validator specification.

jmrog commented:
@jridgewell I think I've addressed all of your comments now, primarily in 8281e8e (which uses the AMP helper methods you mentioned, adds validator rules, and passes the widgetId rather than mutating the iframe element) and 1c19e76 (which doesn't rely on getting title and url from the element's `ownerDocument`, but instead allows the publisher to set the title/url/etc. and uses the `ownerDocument` only as a fallback).

@bpaduch Note that there are some changes to the documentation as a result of the change in 1c19e76.

Thanks!

jridgewell commented:
I'll be out for the next three days. To avoid stalling, I'm forwarding this to @aghassemi.

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,216 @@
+   * @override
+   */
+  preconnectCallback(opt_onLayout) {
+    const {preconnect} = this;
```
too much destructions :) I am normally fine with destructors but `const {preconnect} = this;` is not really readable. let's just keep these as `this.preconnect.`

jmrog replied:
Okay, I agree (I was a little hesitant about this myself). Thanks.

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,216 @@
+   * @override
+   */
+  buildCallback() {
+    const {element} = this;
```
ditto

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,216 @@
+   * @override
+   */
+  renderOutsideViewport() {
+    return true;
```
let's keep normal resource scheduling. I don't see a reason to make an exception for this component.

jmrog replied:
One of the main reasons is this: publishers rely on AddThis tools to provide them with accurate analytics about _visits_ to the pages on which the tools reside. It's fair to say that this is a core use case for our publishers (they like to see data, for example, on things like the number of shares relative to the number of page visits). If the tools don't render unless they are scrolled into view (etc.), then the analytics data provided to publishers will be inaccurate, as not all visits will be registered correctly (since registering the visit using our tools requires at least that the iframe be loaded, which currently happens in`layoutCallback`). Does that make sense?

cramforce replied:
If you'd like to send some analytics ping, the right way to do it is to just emit an image call at the right time. We don't allow iframe based analytics.

I think @zhouyx was working on extension analytics. But something simpler might work here.

aghassemi replied:
@jmrog per @cramforce's comment, you can do:

```
buildCallback() {
 ...
 viewer.whenFirstVisible().then( () => {
   const pixel = new thisn.win.Image();
   pixel.src = 'image-based source for your analytics'
 }
}
```

zhouyx replied:
Extension analytics doesn't apply here. Extension analytics make sure that the analytics is scoped in the extension element, not about the page. 
I think `#buildCallback()` is a good time to send the analytics ping.

jmrog replied:
Hi all, thanks for the input here. Dropping a pixel during the `buildCallback` is a reasonable option for registering a visit to a page; however, does this mean there is no AMP-compliant way to retrieve data from addthis.com cookies client-side in order to send it with the initial visit request? That will cripple some of the analytics that we currently provide to publishers with our standard tools, so I want to be sure.

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,216 @@
+
+    // Fallbacks for url and title
+    if (!this.shareConfig_.url) {
+      this.shareConfig_.url = element.ownerDocument.location.href;
```
`this.getAmpDoc().getUrl();` instead of `element.ownerDocument.location.href;`

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,216 @@
+      this.shareConfig_.url = element.ownerDocument.location.href;
+    }
+    if (!this.shareConfig_.title) {
+      this.shareConfig_.title = element.ownerDocument.title;
```
`this.getAmpDoc().win.document.title`

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,216 @@
+    this.iframe_ = null;
+
+    /** @private {?Promise} */
+    this.iframeLoadPromise_ = null;
```
You  don't need this. It is not used anywhere outside of `layoutCallback`

jmrog replied:
Good catch; thanks.

jmrog commented:
Not really sure what's going on here after that last commit. Now @googlebot seems not to recognize that my company (Oracle) has signed the CLA (despite committing with the same company email address, etc.), and the Travis build seems to have failed due to timeouts unrelated to any of my changes. I suppose I'll check back in tomorrow on those things.

cramforce commented:
Could you explain this a bit more? I don't quite follow the use case. You
could e.g. make a CORS request with cookies and use the response as
configuration for further actions.

On Mon, Oct 2, 2017 at 12:44 PM, Jason Rogers <notifications@github.com>
wrote:

> *@jmrog* commented on this pull request.
> ------------------------------
>
> In extensions/amp-addthis/0.1/amp-addthis.js
> <https://github.com/ampproject/amphtml/pull/11381#discussion_r142235772>:
>
> > +   * @override
> +   */
> +  preconnectCallback(opt_onLayout) {
> +    const {preconnect} = this;
> +    preconnect.url(ORIGIN, opt_onLayout);
> +    preconnect.url('https://m.addthis.com', opt_onLayout);
> +    preconnect.url('https://m.addthisedge.com', opt_onLayout);
> +    preconnect.url('https://cache.addthiscdn.com', opt_onLayout);
> +  }
> +
> +  /**
> +   * @return {boolean}
> +   * @override
> +   */
> +  renderOutsideViewport() {
> +    return true;
>
> Hi all, thanks for the input here. Dropping a pixel during the
> buildCallback is a reasonable option for registering a visit to a page;
> however, does this mean there is no AMP-compliant way to retrieve data from
> addthis.com cookies client-side in order to send it with the initial
> visit request? That will cripple some of the analytics that we currently
> provide to publishers with our standard tools, so I want to be sure.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ampproject/amphtml/pull/11381#discussion_r142235772>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAFeTwnAVn5g299jfuFgoDlMtyB89lJgks5soT0xgaJpZM4Pf9X9>
> .
>

jmrog commented:
@cramforce Ah, you're right, that actually should work. I'll try it out and see how it goes.

jmrog commented:
@cramforce Quick question regarding this:

> You could e.g. make a CORS request with cookies and use the response as configuration for further actions.

Do we have to open the endpoint to CORS, or is using JSONP okay in this case?

cramforce commented:
We cannot support JSONP unfortunately due to CSP restrictions.

jmrog commented:
Hi all,

I'm moving on from AddThis/Oracle to other things, so I've transferred control of my fork/PR over to one of my co-workers at AddThis, @dmvjs. The most recent commit is likely to be my last contribution on this PR (and the tests are expected to fail for this commit, as there are some big changes here -- @dmvjs will be updating the tests as time allows). That said, **please do not review the PR as it currently stands**. Also, please let me know if there is anything else we need to do when it comes to making @dmvjs the point of contact on this going forward.

Thanks!

dmvjs commented:
Greetings, I'm Kirk Elliott aka @dmvjs

I understand that our Oracle Moat team has done some good work with you guys with regards to amp-iframe and their AMP project. Is it possible that we could set up a regular meeting similar to what they are doing to iron out details before we write more code? There are other people on our publisher team (not on github) who I'd like to include in planning and discussions around our approach as well. I'd also like to take a fresh look at our approach and consider using AMP-provided tools where possible. 

If a weekly or biweekly meeting is impractical we can keep discussions here however even a single person-to-person (via Hangout or Zoom) meeting would be a real nice-to-have.

jridgewell commented:
/to @rudygalfi or @ericlindley-g

ericlindley-g commented:
Hi @dmvjs — apologies for the delayed response! Would you be interested in discussing next steps in the public [design review](https://github.com/ampproject/amphtml/blob/master/contributing/design-reviews.md)? After checking with the folks who were working with your team on this, it sounds like the extension is very close to the finish line, and may just need a little bit of discussion to wrap things up. What do you think?

jmrog commented:
> The bad news is that it appears that one or more commits were authored by someone other than the pull request submitter. We need to confirm that all authors are ok with their commits being contributed to this project. Please have them confirm that here in the pull request.

I'm assuming this is a reference to my earlier commits, since @dmvjs has now taken over this PR. On that assumption, I'm ok with my commits being contributed to this project.

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,361 @@
+          });
+
+      // Only the component that registers the page view listens for x-frame events.
+      this.setupListeners_({ampDoc, loc, pubId: this.pubId_});
```
this can't be in the `buildCallback`, should move inside `whenFirstVisible` or `layoutcallback`

Reason is that this method sets up events listeners that end up triggering pixel drops or beacons (any communication that can cause IP leak in prerender mode is forbidden in AMP `buildCallabck` runs in prerender mode)

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,361 @@
+   */
+  createPlaceholderCallback() {
+    const placeholder = createElementWithAttributes(
+        this.win.document,
```
this.element.ownerDocument

aghassemi commented on extensions/amp-addthis/0.1/amp-addthis.js:
```diff
@@ -0,0 +1,361 @@
+    setStyle(placeholder, 'background-color', '#eee');
+
+    const image = createElementWithAttributes(
+        this.win.document,
```
(replace all `this.win.document` with `this.element.ownerDocument`)

aghassemi commented on extensions/amp-addthis/validator-amp-addthis.protoascii:
```diff
@@ -0,0 +1,73 @@
+  attr_lists: "common-extension-attrs"
+}
+
+tags: {  # <amp-addthis>
```
/cc @honeybadgerdontcare

dmvjs replied:
@honeybadgerdontcare @jridgewell @rudygalfi do you have any additional changes or comments while @aghassemi is out on vacation?

honeybadgerdontcare commented on extensions/amp-addthis/validator-amp-addthis.protoascii:
```diff
@@ -0,0 +1,73 @@
+#
+# Copyright 2017 The AMP HTML Authors. All Rights Reserved.
```
replace 2017 with 2018

honeybadgerdontcare commented on extensions/amp-addthis/validator-amp-addthis.protoascii:
```diff
@@ -0,0 +1,73 @@
+  html_format: AMP
+  tag_name: "AMP-ADDTHIS"
+  requires_extension: "amp-addthis"
+  attrs: {
```
please alphabetize the attributes by name.

honeybadgerdontcare commented on extensions/amp-addthis/validator-amp-addthis.protoascii:
```diff
@@ -0,0 +1,73 @@
+  tag_name: "AMP-ADDTHIS"
+  requires_extension: "amp-addthis"
+  attrs: {
+    name: "data-pub-id"
```
Up to you, not required:

If the pub-id follows a certain format then it's possible to restrict the set of values that can be here to follow that format via `value_regex_casei`.

Same with `data-widget-id` below.

aghassemi commented on extensions/amp-addthis/amp-addthis.md:
```diff
@@ -0,0 +1,95 @@
+
+See amp-addthis rules in the AMP validator specification.
+
+## Privacy
```
@dmvjs What this component collects needs from the page and user (e.g. clicks, scrolls, etc) needs to be documented as well.

aghassemi commented:
@honeybadgerdontcare PTAL for validation changes

honeybadgerdontcare commented on extensions/amp-addthis/validator-amp-addthis.protoascii:
```diff
@@ -0,0 +1,73 @@
+# limitations under the license.
+#
+
+tags: {  # amp-addthis
```
We try to keep these tagspecs in the same order and that order typically follows the order that is in the [proto](https://github.com/ampproject/amphtml/blob/master/validator/validator.proto#L385). You can see an [example here](https://github.com/ampproject/amphtml/blob/master/extensions/amp-timeago/validator-amp-timeago.protoascii).

The `html_format` shoudl come first, then `tag_name`, etc.

honeybadgerdontcare commented on extensions/amp-addthis/validator-amp-addthis.protoascii:
```diff
@@ -0,0 +1,73 @@
+    name: "data-pub-id"
+  }
+  attrs: {
+    name: "data-share-description"
```
`data-` attributes are always allowed in AMP documents. So this isn't necessary to list unless you want to add another constraint on it, such as what was done for `data-pub-id` with `mandatory`. If there are no other constraints for this or `data-share-title` then I'd suggest removing them.

dmvjs replied:
👍

aghassemi commented:
@dmvjs Please see above couple of requests for validation rules, we can marge after those are done. Thanks!

aghassemi commented:
@honeybadgerdontcare PTAL, thank you.

rsimha commented on extensions/amp-addthis/0.1/addthis-utils/cuid.js:
```diff
@@ -0,0 +1,80 @@
+  const yearIsSame = date.getFullYear() === now.getFullYear();
+  const monthIsLater = date.getMonth() > now.getMonth();
+  const monthIsSame = date.getMonth() === now.getMonth();
+  const dateIsLater = date.getDay() > now.getDay();
```
@jmrog @honeybadgerdontcare @aghassemi This line is incorrect, since it is comparing the day of the week, and not the date. As written, this logic will break every Saturday.

See https://travis-ci.org/ampproject/amphtml/jobs/372268665#L1472-L1476

It should be replaced with `date.getDate() > now.getDate()`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getDay
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getDate

cramforce replied:
LOL. Please file a bug and lets get the test disabled.

rsimha replied:
Sent out a fix instead. #14960

cramforce commented:
@bpaduch This is unrelated, but I can't see it actually happening. This could have a few causes:

- You are in a captive portal like in an airport wifi
- You have a bad Chrome extension
- Your computer got hacked
- The AMP JS CDN has a major misconfiguration

bpaduch commented:
@cramforce - it appears to be gone now (on corp wifi before/now).  Another extension (amp-beopinion) does continue to have issue (but the creator is [fixing it](https://github.com/ampproject/amphtml/pull/13142#issuecomment-380849448)).

cramforce commented:
@bpaduch While that fix might be legit, it doesn't fix the problem you describe.

What do you get when you go directly to https://cdn.ampproject.org/v0/amp-beopinion-0.1.js
and what is the mime type you see in Chrome DevTools?

This is what I see:

<img width="195" alt="screen shot 2018-04-30 at 11 11 16 am" src="https://user-images.githubusercontent.com/89679/39442780-3918e904-4c67-11e8-8a18-91326518e659.png">

pbakaus commented:
the mime type just happens because it is a 404, and serves an error page :)

jridgewell commented:
It's only released as a canary file. You have to opt-in to get it.

cramforce commented:
Ahh, then this is all WAI!

lmrwork commented:
Thanks!!

pacoesni commented:
Not getting the canonical URL to share on WP.

Ok, here is the thing. I've already implement AddThis extension on AMP for WP template. Everything goes fine except when trying to share. I don't want my users to share /amp/ links. So I get into the custom attributes and try to use like this:

<amp-addthis
    width="320"
    height="70"
    data-pub-id="my_pub_id"
    data-widget-id="widget_id"
    data-share-url="**<?php echo wp_kses_data( get_permalink() ); ?>**">
  </amp-addthis>

I thought that 'data-share-url' would solve my request but it didn't. What I'm missing? 

Check this example: https://www.tekcrispy.com/2018/05/31/fosil-madre-lagartos/amp/

judyn commented:
Having the same issue as @pacoesni, the `data-share-url` shows the correct canonical url in the source but sharing points to the AMP url.
